subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java

```diff
@@ -481,8 +485,18 @@ public void execute(DeleteSpec spec) {
     private void handleCollectedResults(TestCountLogger testCountLogger) {
         if (testCountLogger.hadFailures()) {
             handleTestFailures();
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (shouldFailOnNoMatchingTests()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            } else if (successWithoutTest.getOrNull() == null) {
+                DeprecationLogger.deprecateBehaviour("There is no test to run.")
+                    .withAdvice("Set Test.successWithoutTest to true if you want the task to succeed when there is no test to run.")
+                    .willBecomeAnErrorInGradle9()
+                    .withUpgradeGuideSection(8, "TODO")
```

Comment:
:x: You need to add a section to the upgrade guide and link it here
:lipstick: Also extracting the deprecation logging in a dedicated method would make this clearer.